Kudu.Core/SourceControl/Git/LibGit2SharpRepository.cs

```diff
@@ -0,0 +1,344 @@
+                    using (tracer.Step("Configure git-credential"))
+                    {
+                        FileSystemHelpers.EnsureDirectory(Path.GetDirectoryName(GitCredentialHookPath));
+
+                        string content = @"#!/bin/sh
+if [ " + "\"$1\" = \"get\"" + @" ]; then
+      echo username=dummyUser
+      echo password=dummyPassword
+fi" + "\n";
+
+                        File.WriteAllText(GitCredentialHookPath, content);
+
+                        repo.Config.Set("credential.helper", string.Format("!'{0}'", GitCredentialHookPath));
+                    }
+                }
+                using (tracer.Step("Setup post receive hook"))
+                {
+                    FileSystemHelpers.EnsureDirectory(Path.GetDirectoryName(PostReceiveHookPath));
+
+                    string content = @"#!/bin/sh
+read i
+echo $i > pushinfo
+" + KnownEnvironment.KUDUCOMMAND + "\n";
+
+                    File.WriteAllText(PostReceiveHookPath, content);
+                }
+
+                // NOTE: don't add any new init steps after creating the post receive hook,
+                // as it's also used to mark that Init was fully executed
+            }
+        }
+
+        private string GitCredentialHookPath
+        {
+            get
+            {
+                return Path.Combine(_repositoryPath, ".git", "hooks", "git-credential-invalid.sh");
+            }
+        }
+
+        public override ChangeSet GetChangeSet(string id)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                LibGit2Sharp.Commit commit = repo.Lookup<LibGit2Sharp.Commit>(id);
+
+                if (commit == null)
+                {
+                    return null;
+                }
+
+                return new ChangeSet(commit.Sha, commit.Author.Name, commit.Author.Email, commit.Message, commit.Author.When);
+            }
+        }
+
+        public override void AddFile(string path)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.Stage(path);
+            }
+        }
+
+        public override bool Commit(string message, string authorName, string emailAddress)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                try
+                {
+                    repo.Stage("*");
+                    if (string.IsNullOrEmpty(authorName) ||
+                        string.IsNullOrEmpty(emailAddress))
+                    {
+                        repo.Commit(message);
+                    }
+                    else
+                    {
+                        repo.Commit(message, new Signature(authorName, emailAddress, DateTimeOffset.UtcNow));
+                    }
+                    return true;
+                }
+                catch(LibGit2Sharp.EmptyCommitException)
+                {
+                    return false;
```

Comment:
In this case, maybe the logic could be clearer if

 - Invoke `repo.RetrieveStatus()`
 - Exit early if nothing has changed
 - Stage the list of changes between the index and the workdir that have been detected